[IMP] api: Add and enhance multiple endpoints for improved functionality

JAntonioSalas opened the pull request:
## Summary

This pull request introduces improvements and new features across multiple endpoints to enhance the overall functionality and user experience of the system. The changes include better error handling, more detailed responses, and additional validation.

## Changes Included

1. **/create_schedule_activity_invoice**
    - Added documentation and improved error handling for creating scheduled activities related to invoices.

2. **/get_states_mexico**
    - Enhanced endpoint to retrieve the list of states in Mexico with improved error handling and detailed responses.

3. **/get_inventory_by_sku**
    - Updated endpoint to retrieve inventory details based on SKU, ensuring accurate and detailed information is returned.

4. **/get_inventory**
    - Improved retrieval of inventory details for all products with a SKU, providing more detailed responses.

5. **/send_message_sale_order**
    - Added functionality to post messages to sale orders, with improved logging and error handling.

6. **/create_schedule_activity**
    - Enhanced the creation of scheduled activities related to sale orders, including additional validation and detailed responses.

7. **/confirm_sale_order**
    - Simplified the process of confirming sale orders, with detailed logging and responses.

8. **/send_invoice_by_email/<int:invoice_id>**
    - Improved the functionality for sending invoices by email, ensuring better user feedback and error handling.

9. **/stamp_invoice/<int:invoice_id>**
    - Enhanced the invoice stamping process, including detailed error messages and success responses.

10. **/download_invoice/<int:invoice_id>**
    - Streamlined the invoice download process to generate and deliver PDFs more efficiently.

11. **/get_shipping_info/<model("sale.order"):order>**
    - Added functionality to retrieve detailed shipping information for sale orders, including delivery address details and shipping records.

12. **/register_payment_invoice/<model("account.move"):invoice>**
    - Improved the payment registration process for invoices, ensuring accurate logging and error handling.

13. **/invoice_sale_order**
    - Enhanced the process of invoicing sale orders, including better handling of invoice creation and posting.

14. **/update_sale_order**
    - Improved the update process for sale orders, with better validation and detailed logging.

15. **/create_sale_order**
    - Enhanced the creation process for sale orders, with better validation and handling of optional fields.

16. **/delivery_address**
    - Improved the process for creating or updating delivery addresses, including better handling of address formatting and validation.

17. **/address_invoice**
    - Enhanced the process for creating or updating billing addresses, ensuring better validation and error handling.

18. **/update_contact**
    - Improved the contact update process, with better validation and detailed logging.

19. **/create_contact**
    - Enhanced the contact creation process with better search logic and validation based on email, phone, or store name.

20. **README.rst**
    - README.rst for module documentation

## Related PR
- Closes #24 (close related PR on GitHub)

Please review the changes and provide feedback. Thank you!

JAntonioSalas committed:
[IMP] api: add and enhance several endpoints

Added docstring, clarified the purpose and improved of multiple endpoints to enhance the user experience and maintain consistency in the application.

1. `/create_schedule_activity_invoice`: Improved error handling for creating scheduled activities related to invoices.
2. `/get_states_mexico`: Enhanced the endpoint to retrieve the list of states in Mexico, including better error handling and more detailed responses.
3. `/get_inventory_by_sku`: Updated the endpoint to retrieve inventory details based on SKU, ensuring accurate and detailed inventory information is returned. Filter inventory by location.
4. `/get_inventory`: Improved the retrieval of inventory details for all products with a SKU, adding more detailed responses. Filter inventory by location.
5. `/send_message_sale_order`: Added functionality to post messages to sale orders, with better logging and error handling.
6. `/create_schedule_activity`: Enhanced the creation of scheduled activities related to sale orders, including additional validation and detailed responses.
7. `/confirm_sale_order`: Simplified the process of confirming sale orders and added detailed logging and responses.
8. `/send_invoice_by_email/<int:invoice_id>`: Improved the functionality for sending invoices by email, ensuring better user feedback and error handling.
9. `/stamp_invoice/<int:invoice_id>`: Enhanced the invoice stamping process, including detailed error messages and success responses.
10. `/download_invoice/<int:invoice_id>`: Streamlined the invoice download process to generate and deliver PDFs more efficiently.
11. `/get_shipping_info/<model("sale.order"):order>`: Added functionality to retrieve detailed shipping information for sale orders, including delivery address details and shipping records.
12. `/register_payment_invoice/<model("account.move"):invoice>`: Improved the payment registration process for invoices, ensuring accurate logging and error handling.
13. `/invoice_sale_order`: Enhanced the process of invoicing sale orders, including better handling of invoice creation and posting.
14. `/update_sale_order`: Improved the update process for sale orders, including validation and detailed logging.
15. `/create_sale_order`: Enhanced the creation process for sale orders with better validation and handling of optional fields.
16. Unify scheduling and address endpoints
17. `/delivery_address`: Improved the process for creating or updating delivery addresses, including better handling of address formatting and validation.
18. `/address_invoice`: Enhanced the process for creating or updating billing addresses, ensuring better validation and error handling.
19. `/update_contact`: Improved the contact update process, including better validation and detailed logging.
20. `/create_contact`: Enhanced the contact creation process with better search logic and validation based on email, phone, or store name (contact name).
21. doc: add README.rst for module documentation
22. Filter inventory by location and SKU
23. Unify scheduling and address endpoints

These changes improve the overall robustness and maintainability of the system by ensuring that each endpoint handles data consistently and provides meaningful feedback to users.

Closes #24
Closes #25

hugho-ad commented on hantec_api_ecommerce/__init__.py:
```diff
@@ -0,0 +1,2 @@
+
+from . import controllers
```
EOF

hugho-ad commented on hantec_api_ecommerce/__manifest__.py:
```diff
@@ -0,0 +1,12 @@
+    "installable": True,
+    "development_status": "Production/Stable",
+    "maintainers": ["C&O PROJECTS AND SOLUTIONS"],
+}
```
EOF

hugho-ad commented on hantec_api_ecommerce/controllers/__init__.py:
```diff
@@ -0,0 +1,2 @@
+
+from . import main
```
EOF

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+class MainController(Controller):
+    @route("/create_contact", methods=["POST"], type="json", auth="user")
+    def create_contact(self):
+        """
```
(docstring-first-line-empty) First line empty in method docstring

move the line below in the same line please

```suggestion
        """Combines ...
```

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+
+    @route("/update_contact", methods=["POST"], type="json", auth="user")
+    def update_contact(self):
+        """
```
apply for all docstrings

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+        if store_name:
+            store_name_suffix = store_name[len(store_name) - 4 :]
+
+            domain = [("name", "ilike", f"%{store_name_suffix}")]
```
```suggestion
            domain = [("name", "=", f"%{store_name_suffix}")]
```
why not to use = operator?

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+
+        """
+        # Find the invoice by its ID
+        invoice = request.env["account.move"].browse(invoice_id)
```
```suggestion
        invoice = request.env["account.move"].browse(invoice_id).exists()
```

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+
+        """
+        data = request.jsonrequest
+        sale_order_id = data.get("sale_order_id")
```
I  would rather avoid this, you can use <model("sale.order"):order>

that way you deacrease code lines, and automatically you got the order id object already as an browseable record set

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+        }
+
+    @route("/get_states_mexico", methods=["GET"], type="json", auth="user")
+    def get_states_mexico(self):
```
I think you can retrive all, and filter them at django

similar to 
https://github.com/odoo/odoo/blob/ab01f91913f32e11cc20287f719933849f0949eb/addons/website/controllers/main.py#L215-L218

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+        """
+        sku = request.jsonrequest.get("sku")
+        product = request.env["product.product"].search([("default_code", "=", sku)])
+        product_fields = ["name", "default_code", "qty_available", "virtual_available"]
```
it's important to know in which location you need to know the inventory values

don you have a specific location for ecommerce demands?

JAntonioSalas replied:
Yes, there are several locations

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+            "message": f"Message successfully posted in sale order with ID: {sale_order_id}."
+        }
+
+    @route("/get_inventory", methods=["GET"], type="json", auth="user")
```
merge this controller with the one below

just add another route in for the same method

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+        user_id = data.get("user_id", request.env.uid)  # User assigned to the activity
+
+        # Create the scheduled activity
+        activity = request.env["mail.activity"].create(
```
no man, use invoice.activity_schedule
https://github.com/odoo/odoo/blob/ab01f91913f32e11cc20287f719933849f0949eb/addons/mail/wizard/mail_activity_schedule.py#L216

hugho-ad replied:
@JAntonioSalas 
missing this

JAntonioSalas replied:
I already pushed this change and the others @hugho-ad

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+            }
+        )
+        logger.info(
+            "Scheduled activity created with ID %s for invoice %s",
```
do you really need this log?

what about move it as logger.debug

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+        """
+        # Delivery address data
+        shipping_address = order.partner_shipping_id
+        address_data = {
```
better use 
shipping_address.read([fields])

hugho-ad replied:
this

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,915 @@
+            ),
+        }
+
+        # Get related shipping records
```
what about to move this logic to the model picking or order itself?

you will need to createa dir 
models/sale_order.py

and inside the sale_order.py you create the get method with all this logic, this is in order to separated the controllers and the data models

hugho-ad replied:
this

JAntonioSalas commented:
I made the changes, could you please review it @hugho-ad ?

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,799 @@
+
+        inventory_data = []
+        for product in products:
+            quants = request.env["stock.quant"].search([("product_id", "=", product.id), ("location_id", "=", location_id)])
```
```suggestion
         product.with_context(location=location_id).read(["name", "default_code", "qty_available", "virtual_available"])
```
Don't need to look for quants

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,799 @@
+        contact_data = data.get("contact_data", {})
+
+        if email or phone:
+            email_pattern = r"^([a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*)"  # Email pattern to get the value before @
```
line too long (140 > 119 characters)

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,799 @@
+        partner_id = data.get("partner_id")
+        invoice_data = data.get("invoice_data")
+
+        partner = env["res.partner"].browse(partner_id)
```
not used

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,799 @@
+        partner_id = data.get("partner_id")
+        delivery_data = data.get("address_data")
+        only_create = data.get("only_create", False)  # Boolean to omit address update
+        partner = env["res.partner"].browse(partner_id)
```
same, not used

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,799 @@
+
+        """
+        # Find the invoice
+        invoice = request.env["account.move"].browse(invoice_id).exists()
```
not used

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,799 @@
+        return order.get_shipping_info()
+
+    @route(
+        "/download_invoice/<int:invoice_id>", methods=["GET"], type="http", auth="user"
```
what about to use the model
https://github.com/Vauxoo/addons-vauxoo/blob/15.0/website_sale_brand_catalogs/controllers/portal.py#L10

JAntonioSalas replied:
Do I use this model for a new endpoint? @hugho-ad

hugho-ad replied:
for the actual endpoints, the Idea is that you avoid make always the browse over the id you get from data

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,799 @@
+        Returns:
+            dict: A dictionary with a message and the inventory details of the products.
+        """
+        location_id = request.jsonrequest.get("location_id") or request.params.get("location_id")
```
not used location_id

JAntonioSalas commented:
I pushed the new changes, could you please review it ? @hugho-ad

hugho-ad commented on hantec_api_ecommerce/models/sale_order.py:
```diff
@@ -0,0 +1,69 @@
+from odoo import models, fields
```
```suggestion
from odoo import models
```
not used

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,817 @@
+        }
+
+    @route("/delivery_address", methods=["POST"], type="json", auth="user")
+    def delivery_address(self):
```
suggestion:

you could merge this endpoint at the one above at create_address_invoice

it's the same code, just here are handaling the only_create boolean

my recomendation is to merge them in a sigle method

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,817 @@
+        type="json",
+        auth="user",
+    )
+    def create_schedule_activity_invoice(self, invoice=False):
```
same here, you can merge this method with the above, you can handle them if you do it like this
https://github.com/odoo/odoo/blob/15.0/addons/website_crm_partner_assign/controllers/main.py#L210-L214

se how they handle different records sets by model type

the thing is you can avoid duplicated code

JAntonioSalas commented:
I pushed the new changes @hugho-ad

JAntonioSalas commented:
could you please review it? @hugho-ad

hugho-ad commented on hantec_api_ecommerce/controllers/main.py:
```diff
@@ -0,0 +1,724 @@
+
+        if only_create:
+            address = False
+        if address:
```
```suggestion
        if address and not only_create:
```

this way we can avoid the if above

hugho-ad commented:
LGTM :+1: 

Just squash again please

git rebase --autosquash -i origin/15.0

JAntonioSalas commented:
It is ready @hugho-ad

hugho-ad commented:
@JAntonioSalas 

one important thing, its the uni tests, did you plan to include them?

JAntonioSalas commented:
I think not at the moment because they need to use those endpoints as soon as possible. @hugho-ad

JAntonioSalas committed:
[ADD] ks_dashboard_ninja: module to dashboard with AI

JAntonioSalas committed:
[ADD] base_user_role: module to extend the functionality regarding users and groups management.

JAntonioSalas committed:
[IMP] hantec_api_ecommerce: Endpoint to get product ID by internal reference

JAntonioSalas committed:
[IMP] API: Search by name and create contact

JAntonioSalas committed:
[ADD] eg_lock_price:  New module Lock Price
This app will help the manager to set the price fix as the product price in sale order, purchase order, invoicing and account price.

JAntonioSalas committed:
[ADD] sale_product_stock_app:  New module sale product stock

JAntonioSalas committed:
[IMP] sale_product_stock_app:  Add values in SaleWarehouseLine

JAntonioSalas committed:
[IMP] sale_product_stock_app:  Change position fields and string of fields

JAntonioSalas committed:
[IMP] sale_product_stock_app:  Translate tags

JAntonioSalas committed:
[IMP] hantec_api_ecommerce: new fields
Add new optional fields when a sale order is created

JAntonioSalas committed:
[IMP] hantec_api_ecommerce: new field
Add new optional field when a sale order is created

JAntonioSalas committed:
[REF] hantec_api_ecommerce: Remove optional fields.
Add dynamic optional fields when creating a sales order

JAntonioSalas committed:
[IMP] hantec_api_ecommerce: Remove required field
Remove required field when creating sale order.

hugho-ad committed:
[ADD] .gitmodules: Welcome Server Backend and Base_user_role module

JAntonioSalas committed:
Merge 15.0-add-user-role-module-hugho-ad into ecommerce

JAntonioSalas committed:
[IMP] hantec_api_ecommerce: Allow more fields in invoices.

JAntonioSalas committed:
[IMP] hantec_api_ecommerce: Add new field in invoices.

JAntonioSalas committed:
[IMP] hantec_api_ecommerce: Add field in invoices if exists.

JAntonioSalas committed:
[MERGE] ecommerce: Get last version

JAntonioSalas committed:
[FIX] sale_order: invalid field on model sale order

JAntonioSalas committed:
[FIX] gitmodules: remove module

JAntonioSalas committed:
[REL] ecommerce: update versions to 18

JAntonioSalas committed:
Trigger update

JAntonioSalas committed:
[FIX]: hantec: update version manifest

JAntonioSalas committed:
[ADD] server-backend: add module to gitmodules

JAntonioSalas committed:
[FIX] gitmodules

JAntonioSalas committed:
[FIX] gitmodules: branch version

JAntonioSalas committed:
[REM] gitmodules: remove submodule

JAntonioSalas committed:
Trigger update

JAntonioSalas committed:
Trigger update

JAntonioSalas committed:
[REV] gitmodules: reverting modules version

JAntonioSalas committed:
Trigger update

JAntonioSalas committed:
[FIX] submodules: change branch to 17

JAntonioSalas committed:
[FIX] submodules: change branch to 15
